bpf: Split bpf_lxc CT lookups to their own tail calls

pchaigno opened the pull request:
### Summary

This pull request moves the conntrack lookups of bpf_lxc to their own BPF programs via tail calls, in order to reduce complexity and program size significantly.

Commits 1–5 refactor the code in preparation for the main changes. Commit 6 introduces the new tail calls and programs. Commits 7 and 8 move the CT lookups to the new BPF programs. Commit 9 handles a special case for older kernels where we can't pass map values to map helpers.

### Results

This pull request results in a strong reduction of complexity for the largest BPF programs in bpf_lxc. I measured it on three kernels (4.19 should be the same), comparing b5fd71b9f873942764e90e41b39821ea48152832 (master) and this branch (split), and kept only the four largest programs.

  | 4.9 | 4.9 | 5.4 | 5.4 | netnext | netnext
-- | -- | -- | -- | -- | -- | --
**Program** | **master** | **split** | **master** | **split** | **master** | **split**
1st | 87518 | 47746 | 95082 | 61527 | 218471 | 109460
2nd | 64651 | 42614 | 61527 | 46953 | 173102 | 89631
3rd | 26806 | 27388 | 57546 | 24660 | 109460 | 59405
4th | 26409 | 24833 | 54030 | 22002 | 89631 | 48886
Sum | 205384 | 142581 | 268185 | 155142 | 590664 | 307382

Some of the largest programs are not affected because they are NAT handling logic and this pull request doesn't touch them. That said, the programs we have the most complexity issues with are those affected.

To get a visual impression, here is a bar chart of the sums of the four largest programs (last row from table):

![ct-split-complexity-reduction](https://user-images.githubusercontent.com/1764210/170317907-50eb4d76-bca6-4f57-9975-fcd3a9a6a5c2.png)

pchaigno committed:
bpf: Don't pass the full connection tuple unless needed

This commit updates lb4_ctx_restore_state's prototype to take the
destination address directly instead of the full connection tuple. Only
the destination address is needed so best to make this clear for the
caller.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Move CT tuple initialization just before CT lookup

The CT tuple isn't needed before the CT lookup happens. In a subsequent
commit, both the CT tuple initialization and the CT lookup will happen
in a previous BPF program (via tail calls). By moving those two pieces
together, we simplify the subsequent commit where we will have to move
the whole thing around.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Initialize csum offset only when needed

This may help reduce complexity because in IPv4's case it means one
branch moves lower in the control-flow graph (so it should reduce the
total number of paths through the program). Assuming LLVM doesn't
rewrite all that.

It also simplifies a subsequent commit where the CT lookups are moved
to tail called programs.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Make union v6addr const where possible

This is mostly to clarify things for callers, but it will also slightly
simplify a subsequent commit.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Copy map value to stack on <4.18 kernels

Before Linux commit d71962f3e627 (4.18), map helpers were not allowed to
access map values directly. So for those older kernels, we need to copy
the map value from the tail call buffer to the stack first. We don't have
a probe for that, but the bpf_fib_lookup helper was introduced in the
same Linux release. So let's use that helper until we have a proper
probe.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Introduce tail call for CT lookups

This commit introduces new tail call programs for the IPv4 and IPv6
conntrack lookups.

More specifically, it will add the two programs below in the chains of
BPF programs (in the case where all programs are compiled in):

      from-container or CILIUM_MAP_EGRESSPOLICY/LXC_ID
      CILIUM_CALL_IPV{4,6}_FROM_LXC
    + CILIUM_CALL_IPV{4,6}_CT_EGRESS
      CILIUM_CALL_IPV{4,6}_FROM_LXC_CONT

      to-container
    + CILIUM_CALL_IPV{4,6}_CT_INGRESS
      CILIUM_CALL_IPV{4,6}_TO_ENDPOINT

      CILIUM_MAP_POLICY/LXC_ID
    + CILIUM_CALL_IPV{4,6}_CT_INGRESS
      CILIUM_CALL_IPV{4,6}_CT_INGRESS_POLICY_ONLY

The programs are currently noop and will be written in the subsequent
commit. We rely on two macros for IPv4 and IPv6 because the programs
will use different structures depending on the IP family.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Split CT lookup to its own tail call

In this commit we will move the conntrack lookups of bpf_lxc to the BPF
programs created in the previous commit. To pass data between the CT
lookup BPF programs and their consumers, we will use two new per-CPU
BPF arrays of one element each, called "tail call buffers". They will
hold the tuple (IP-family specific), the looked up CT state, the monitor
value and the return code from the CT lookup.

We have to pass the CT tuple as part of the tail call buffers because
it is also populated (L4 fields) during the CT lookup.

Finally, if either the tail call buffer updates or the lookups fail, we
will drop the packet with error code DROP_INVALID_TC_BUFFER. Note the
lookups should never fail.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Refer to ct_state and tuple by pointer

In the subsequent commit, ct_state and tuple will be read from a BPF map
passed by a previous BPF program through tail call. They will therefore
be referred to by pointer. That implies changing all accesses to those
variables, in a lot of different places.

In an attempt to ease reviews, the present commit simply changes
ct_state and tuple to pointer types. This commit therefore has no
functional changes (even the bytecode is likely unchanged) and the next
commit is hopefully a bit clearer with only the functional changes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -1217,12 +1297,12 @@ ipv6_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label,
-	struct ipv6_ct_tuple tuple_on_stack, *tuple = &tuple_on_stack;
-	struct ct_state ct_state_new = {};
+	struct ct_state *ct_state, ct_state_new = {};
+	int ret, l4_off, verdict, hdrlen, zero = 0;
```
Move the definition for `l4_off` down to its only user?

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -1524,35 +1669,30 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
-	tuple->saddr = ip4->saddr;
 	orig_sip = ip4->saddr;
 
 	l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
```
Keep things in sync with `ipv6_policy()`, and move `l4_off` down into the only branch where it's used?

julianwiedmann reviewed (commented):
Still need to dig into the tail-call details, but the preceding cleanups are sweet already. Thanks!

One minor nit-pick - having `bpf: Replace l3_off by its constant value` earlier in the series would save a tiny bit of churn when looking at `bpf: Move CT tuple initialization just before CT lookup`.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -275,6 +281,12 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+		if (hdrlen < 0)
+			return hdrlen;
+
+		l4_off = ETH_HLEN + hdrlen;
```
We could pass up the `l4_off` from the CT lookup. But presumably not worth it.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -173,26 +188,17 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
-	/* Pass all outgoing packets through conntrack. This will create an
-	 * entry to allow reverse packets and return set cb[CB_POLICY] to
-	 * POLICY_SKIP if the packet is a reply packet to an existing incoming
-	 * connection.
```
Is there any information in this comment that's worth keeping?

pchaigno replied:
I don't think so. It's not even correct anymore.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -195,8 +195,21 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+	 * We don't have a probe for that, but the bpf_fib_lookup helper was
+	 * introduced in the same release.
+	 */
+#if HAVE_PROG_TYPE_HELPER(sched_cls, bpf_fib_lookup)
```
Can we hide this behind some `HAVE_DIRECT_ACCESS_TO_MAP_VALUES` (that expands to `HAVE_PROG_TYPE_HELPER(...)`), so that we don't need this comment everywhere?

julianwiedmann reviewed (commented):
Some comments (rather code-layout related), otherwise LGTM. And the gains are awesome :rocket:.

ti-mo commented on bpf/bpf_lxc.c:
```diff
@@ -101,6 +101,21 @@ encode_custom_prog_meta(struct __ctx_buff *ctx, int ret, __u32 identity)
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, __u32);
+	__type(value, struct ct_buffer6);
+	__uint(pinning, LIBBPF_PIN_BY_NAME);
```
Does this need to be pinned? It's not a prog array, so it should follow the normal lifecycle of bpf progs.

pchaigno replied:
Good call. I removed it.

pchaigno committed:
bpf: Replace l3_off by its constant value

Variable l3_off is actually always a constant so we can just use the
constant directly. That makes it clear to the reader that it won't
change.

Signed-off-by: Paul Chaignon <paul@cilium.io>

julianwiedmann reviewed (approved):
Looks good, thanks!
